### PR TITLE
service and api handlers are created

### PR DESCRIPTION
**Pull request by ahsanfayaz52**

### CODE REVIEW

File: db/mongo/client.go [hunk 1]
```diff
@@ -47,47 +46,46 @@ func (c client) SaveEmployee(employee *models.Employee) (string, error) {
 	opts := options.Update().SetUpsert(true)
 	collection := c.conn.Database(viper.GetString(config.DbName)).Collection(collectionName)
 	if _, err := collection.UpdateOne(context.TODO(), bson.M{"id": employee.ID}, bson.M{"$set": employee}, opts); err != nil {
-		panic("oops cannot update employee")
+		return "", errors.Wrap(err, "failed to add/update employee")
 	}
-	return "", nil
+	return employee.ID, nil
```
Comment (ahab94) [inline db/mongo/client.go:51]: Weren't we covering this with unit tests?

File: db/mongo/client.go [hunk 2]
```diff
@@ -47,47 +46,46 @@ func (c client) SaveEmployee(employee *models.Employee) (string, error) {
...
 }
 
 func (c client) GetEmployeeByID(id string) (*models.Employee, error) {
 	var employee *models.Employee
 	collection := c.conn.Database(viper.GetString(config.DbName)).Collection(collectionName)
 	if err := collection.FindOne(context.TODO(), bson.M{"id": id}).Decode(&employee); err != nil {
 		if errors.Is(err, mongo.ErrNoDocuments) {
-			panic("oops error occurred ")
```
Comment (ahab94) [inline db/mongo/client.go:60]: panic should never be used anywhere except in main.

File: db/mongo/client.go [hunk 3]
```diff
@@ -47,47 +46,46 @@ func (c client) SaveEmployee(employee *models.Employee) (string, error) {
...
+			return nil, errors.Wrap(err, "failed to get employee")
 		}
-		return nil, err
 	}
 	return employee, nil
 }
 
-func (c client) DeleteEmployee(id string) error {
+func (c client) DeleteEmployee(id string) (string, error) {
```
Comment (ahab94) [inline db/mongo/client.go]: this should only return error.

File: db/mongo/client.go [hunk 4]
```diff
@@ -47,47 +46,46 @@ func (c client) SaveEmployee(employee *models.Employee) (string, error) {
...
 	if err != nil {
-		log().Info("oops no records fetched")
+		return nil, errors.Wrap(err, "failed to retrieve employees")
 	}
 	defer cursor.Close(context.TODO())
 	for cursor.Next(context.TODO()) {
+		var em *models.Employee
 		if err = cursor.Decode(&em); err != nil {
-			log().Info("cannot add elements to list")
+			return nil, errors.Wrap(err, "oops error occurred while scanning.....")
```
Comment (ahab94) [inline db/mongo/client.go]: lets not use the word `oops`. and `.....`

File: db/mongo/client.go [hunk 5]
```diff
@@ -47,47 +46,46 @@ func (c client) SaveEmployee(employee *models.Employee) (string, error) {
...
 	}
-	return nil
+	return "employee is removed successfully", nil
 }
 
-func (c client) ListEmployee() (*[]models.Employee, error) {
-	var employee []models.Employee
-	var em models.Employee
+func (c client) ListEmployee() ([]*models.Employee, error) {
+	var employee []*models.Employee
```
Comment (ahab94) [inline db/mongo/client.go:76]: should be employees

File: handlers/list_employees.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	if err != nil {
+		switch apiErr := err.(*domainErr.APIError); {
+		case apiErr.IsError(domainErr.NotFound):
+			return operations.NewGetEmployeeByIDNotFound()
+		default:
+			return operations.NewGetEmployeeByIDInternalServerError()
+		}
+	}
+	var emList []*models.Employee
+	i := 0
```
Comment (ahab94) [inline handlers/list_employees.go]: Just range over the slice... much simpler.

**Review by ahab94 — CHANGES_REQUESTED** [review]
I am seeing you fixed a lot of issues.. but were test cases working properly?

File: db/mongo/client.go (hunk 1 above)
Comment (ahsanfayaz52) [reply db/mongo/client.go:51]: yes sir

File: db/mongo/client.go (hunk 2 above)
Comment (ahsanfayaz52) [reply db/mongo/client.go:60]: got it sir, i removed it

File: db/mongo/client.go (hunk 3 above)
Comment (ahsanfayaz52) [reply db/mongo/client.go]: done

File: db/mongo/client.go (hunk 4 above)
Comment (ahsanfayaz52) [reply db/mongo/client.go]: done

File: db/mongo/client.go (hunk 5 above)
Comment (ahsanfayaz52) [reply db/mongo/client.go:76]: done

File: handlers/list_employees.go (hunk 6 above)
Comment (ahsanfayaz52) [reply handlers/list_employees.go]: done

File: db/mongo/client.go (hunk 1 above)
Comment (ahsanfayaz52) [reply db/mongo/client.go:51]: we are not covering employee.ID in unit tests

**Pull request merged**